src/libponyrt/sched/scheduler.c

```diff
@@ -333,16 +394,64 @@ static pony_actor_t* steal(scheduler_t* sched)
     //    stealing for generating far fewer block/unblock messages.
     if (!block_sent)
     {
-      if (steal_attempts < scheduler_count)
+      uint32_t current_active_scheduler_count = get_active_scheduler_count();
+      if (steal_attempts < current_active_scheduler_count)
       {
         steal_attempts++;
       }
-      else if ((!sched->asio_noisy) &&
-        ((tsc2 - tsc) > 1000000) &&
+      else if (((tsc2 - tsc) > 1000000) &&
         (ponyint_mutemap_size(&sched->mute_mapping) == 0))
       {
-        send_msg(0, SCHED_BLOCK, 0);
-        block_sent = true;
+        bool cmp_val = false;
+
+        // if we're the highest active scheduler thread
+        // and there are more active schedulers than the minimum requested
+        if ((sched == &scheduler[current_active_scheduler_count - 1])
+          && (current_active_scheduler_count > min_scheduler_count) &&
+          atomic_compare_exchange_strong_explicit(&scheduler_count_changing,
+          &cmp_val, true, memory_order_relaxed, memory_order_relaxed))
+
+        {
+          // let sched 0 know we're suspending
+          send_msg(0, SCHED_SUSPEND, 0);
+
+          // decrement active_scheduler_count so other schedulers know we're
+          // sleeping
+          atomic_fetch_sub_explicit(&active_scheduler_count, 1,
+            memory_order_relaxed);
+
+          cmp_val = true;
+
+          // unlock the bool that controls modifying the active scheduler count
+          // variable
+          atomic_compare_exchange_strong_explicit(&scheduler_count_changing,
+            &cmp_val, false, memory_order_relaxed, memory_order_relaxed);
+
+          // sleep waiting for signal to wake up again
+#if defined(PLATFORM_IS_WINDOWS)
+          ponyint_thread_suspend(sched->wait_event_object);
+#else
+          ponyint_thread_suspend(PONY_SCHED_SLEEP_WAKE_SIGNAL);
```

Comment:
It seems to me that there is a race condition here, with the variable modification and thread suspend not being a single atomic operation. For instance, with the following sequence:

```
T1: unlock (count_changing is false)
T2: in sched_maybe_wakeup: lock, read, signal wakeup to T1
T1: suspend
```

If Windows events and pthread signals aren't queued up if sent when the receiver isn't waiting for them, then this is a bug.

Reply:
Hmm.. Good point about the potential race condition here.

For Windows, `SetEvent` documentation (https://msdn.microsoft.com/en-us/library/windows/desktop/ms686211(v=vs.85).aspx) states:

> The state of an auto-reset event object remains signaled until a single waiting thread is released, at which time the system automatically sets the state to nonsignaled. If no threads are waiting, the event object's state remains signaled.

Based on the above quote, I don't think the race condition exists on Windows and if the event gets signalled by `T2` before `T1` suspends, `T1` should noticed it is signalled immediately and return right away due to the event remaining signalled until at least one thread is woken up by it.

For Posix platforms, I've made a change to `block` the signal for waking threads by default for all threads. The note at https://notes.shichao.io/apue/ch10/#reliable-signal-terminology-and-semantics states:

> A process has the option of blocking the delivery of a signal. If a signal that is blocked is generated for a process, and if the action for that signal is either the default action or to catch the signal, then the signal remains pending for the process until the process either:
> * unblocks the signal, or
> * changes the action to ignore the signal.

Based on the above quote, I don't think the race condition exists on Posix platforms (as long as the signal is properly blocked) and if `T1` gets signalled by `T2` before `T1` suspends, `T1` should noticed it is signalled immediately and return right away due to the signal remaining in `pending` state until the thread unblocks it for the `sigwait`.